torch/csrc/autograd/custom_function.cpp

```diff
@@ -26,6 +29,149 @@ Variable VariableInfo::zeros(at::OptionalDeviceGuard& device_guard) const {
   }
 }
 
+void _foreach_tensor(
+    const std::function<void (size_t, size_t, const at::Tensor&)>& fn,
+    torch::jit::Stack* stack,
+    size_t stack_start,
+    size_t size) {
+  // Enumerate over tensors in a stack, including ones in TensorLists
+  int idx = 0;
+  for (const auto i : c10::irange(size)) {
+    auto& ivalue = (*stack)[stack_start + i];
+    if (ivalue.isTensor()) {  // true for optional tensor that has value
+      const auto& tensor = ivalue.toTensor();
+      fn(idx, i, tensor);
+      idx++;
+    } if (ivalue.isTensorList()) {
+      for (const auto& tensor : ivalue.toTensorVector()) {
```

Comment:
no don't do this, this will allocate a std::vector and then iterate over it, but the intermediate allocation is completely unnecessary. Use toListRef() instead; see aten/src/ATen/core/PythonFallbackKernel.cpp for an example